### PR TITLE
Make it easy to configure ssl

### PR DESCRIPTION
**Pull request by lburgazzoli**
This PR lacks of tests, any help on properly setting up an
etcd instance with ssl would be really welcome

### CODE REVIEW

File: src/main/java/com/coreos/jetcd/ClientBuilder.java [hunk 1]
```diff
@@ -145,6 +148,20 @@ public ClientBuilder setLazyInitialization(boolean lazyInitialization) {
...
+    return sslContext;
+  }
+
+  /**
+   * SSL/TLS context to use instead of the system default. It must have been configured with {@link
+   * GrpcSslContexts}, but options could have been overridden.
+   *
+   * @param sslContext the ssl context
+   */
+  public void setSslContext(SslContext sslContext) {
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/ClientBuilder.java]: java still prefer ssl than tls?

**Comment by fanminshi** [comment]
@lburgazzoli the current integration test platform is not good. It needs to be worked on. In the meantime, could you some sort of manual test? And then post the manual test result to this pr.

File: src/main/java/com/coreos/jetcd/resolver/SimpleEtcdNameResolverFactory.java [hunk 2]
```diff
@@ -46,4 +49,24 @@ public String getDefaultScheme() {
   public String name() {
     return NAME;
   }
+
+  public static SimpleEtcdNameResolverFactory forEndpoints(String... endpoints) {
```
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/resolver/SimpleEtcdNameResolverFactory.java]: prefer not to have `Etcd` keyword in naming unless you think that it's more clear.

File: src/main/java/com/coreos/jetcd/resolver/SimpleEtcdNameResolverFactory.java [hunk 3]
```diff
@@ -46,4 +49,24 @@ public String getDefaultScheme() {
...
+    return forEndpoints(Arrays.asList(endpoints));
+  }
+
+  public static SimpleEtcdNameResolverFactory forEndpoints(List<String> endpoints) {
```
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/resolver/SimpleEtcdNameResolverFactory.java]: prefer not to have `Etcd` keyword in naming unless you think that it's more clear.

File: src/main/java/com/coreos/jetcd/ClientBuilder.java (hunk 1 above)
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/ClientBuilder.java]: @xiang90 it seems so. if you look the grpc/auth guide for java; "ssl" is used instead of "tls" in naming. https://grpc.io/docs/guides/auth.html#with-server-authentication-ssltls-4

File: src/main/java/com/coreos/jetcd/resolver/SimpleEtcdNameResolverFactory.java (hunk 2 above)
Comment (lburgazzoli) [inline src/main/java/com/coreos/jetcd/resolver/SimpleEtcdNameResolverFactory.java]: I will change it but I'm not the author of the class :)

Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/resolver/SimpleEtcdNameResolverFactory.java]: oh i see. My bad. don't change it. probably do it in a different pr.

File: src/main/java/com/coreos/jetcd/resolver/SimpleEtcdNameResolverFactory.java (hunk 3 above)
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/resolver/SimpleEtcdNameResolverFactory.java]: nvm.

**Comment by lburgazzoli** [comment]
I will try to add some tests as soon as possible but in the meantime it would be nice to merge this PR so people can eventually test it and report issues.

**Comment by fanminshi** [comment]
@lburgazzoli still reviewing the code; need to understand how tls works in grpc.

**Comment by lburgazzoli** [comment]
In the meantime I've renamed the class, not an issue to change it in this PR IMHO as lots of things are still moving ;)

File: src/main/java/com/coreos/jetcd/ClientBuilder.java (hunk 1 above)
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/ClientBuilder.java]: ok. good to know.

**Comment by xiang90** [comment]
can we add a test for this? or please create an issue to track ssl testing.

**Comment by lburgazzoli** [comment]
@xiang90 I've opened the issue #162

File: src/main/java/com/coreos/jetcd/resolver/SimpleNameResolverFactory.java [hunk 4]
```diff
@@ -46,4 +49,24 @@ public String getDefaultScheme() {
...
+          endpoint = "http://" + endpoint;
+        }
+        return new URI(endpoint);
+      } catch (URISyntaxException e) {
+        throw new IllegalArgumentException(e);
+      }
+    }).collect(Collectors.toList());
+
+    return new SimpleNameResolverFactory(uris);
+
```
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/resolver/SimpleNameResolverFactory.java]: no extra space after return?

File: src/main/java/com/coreos/jetcd/internal/impl/ClientConnectionManager.java [hunk 5]
```diff
@@ -136,4 +146,87 @@ synchronized void close() {
...
+    channelBuilder.nameResolverFactory(factory);
+
+    if (builder.getSslContext() != null) {
+      channelBuilder.sslContext(builder.getSslContext());
+    } else {
+      channelBuilder.usePlaintext(true);
+    }
+
+    return channelBuilder;
+
```
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/internal/impl/ClientConnectionManager.java]: no extra space after return?

File: pom.xml [hunk 6]
```diff
@@ -109,6 +110,13 @@
           <version>${grpc.version}</version>
           <scope>test</scope>
         </dependency>
+        <dependency>
+          <groupId>io.netty</groupId>
+          <artifactId>netty-tcnative-boringssl-static</artifactId>
+          <version>${netty-tcnative.version}</version>
+          <classifier>${os.detected.classifier}</classifier>
+          <scope>test</scope>
```
Comment (lburgazzoli) [inline pom.xml]: Is is indeed for test only as jetcd user should be able to use different versions (i.e. dynamic linked) so netty-tcnative should not be a transitive dependency

File: src/main/java/com/coreos/jetcd/internal/impl/ClientConnectionManager.java (hunk 5 above)
Comment (lburgazzoli) [inline src/main/java/com/coreos/jetcd/internal/impl/ClientConnectionManager.java]: I'd recommend to improve checkstyle to include any style check you'd need.

File: pom.xml (hunk 6 above)
Comment (fanminshi) [inline pom.xml]: Suppose that I export jetcd as jetcd.jar. If the user wants to use jetcd.jar, then the user also needs include a ` netty-tcnative` pkg in order to build the code with jetcd.jar properly? since jetcd needs to be dynamically linked to a `netty-tcnative` pkg. Is my understand of `dynamic linked` in for java jar correct?

Comment (lburgazzoli) [inline pom.xml]: no netty-tcnatyve is not required to run jetcd, it is an optional dependency if you want to use a fast ssl implementation based on openssl.

Comment (fanminshi) [inline pom.xml]: Sorry I am not an export on maven dependency. More question, If I scope the `netty-tcnative-boringssl-static` to test. then the compiled scope of jetcd will not have `io.netty.handler.ssl.SslContext` defined anywhere. this means that the user must import another jar that has `io.netty.handler.ssl.SslContext`?

Comment (lburgazzoli) [inline pom.xml]: SslContext is not part of netty-tcnative which contains only the native binding (via jni) for openssl so the implementation to use is chosen at runtime according to the classpath.

Comment (fanminshi) [inline pom.xml]: I think I know what's going on. `Netty transport` used by jetcd's grpc needs a mechanism to do TLS. The `netty-tcnative-boringssl-static` includes `BoringSSL` where grpc can use it to establish secure connection. On the other hand, user can also set up `Openssl` locally and use `netty-tcnative` to dynamically link to the local `Openssl` binary to establish secured connection. Is that what's going on?

I think we might need to add doc on how to properly tls on jetcd.  This link is helpful https://github.com/grpc/grpc-java/blob/master/SECURITY.md

Comment (lburgazzoli) [inline pom.xml]: Correct, I've added netty-tcnative-boringssl-static to test classpath so tests (when implemented) will use the same ssl implementation regardless of the OS

**Comment by fanminshi** [comment]
lgtm

**Comment by fanminshi** [comment]
could you change commit title to
`client: make it easy to configure ssl`? thanks.

**Comment by lburgazzoli** [comment]
done

**Comment by fanminshi** [comment]
Thanks for the pr.
